operator/controllers/keptnevaluation/providers/dynatrace.go

```diff
@@ -0,0 +1,116 @@
+package providers
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/go-logr/logr"
+	klcv1alpha1 "github.com/keptn/lifecycle-toolkit/operator/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type KeptnDynatraceProvider struct {
+	Log       logr.Logger
+	client    http.Client
+	k8sClient client.Client
+}
+
+type DynatraceResponse struct {
+	TotalCount int               `json:"totalCount"`
+	Resolution string            `json:"resolution"`
+	Result     []DynatraceResult `json:"result"`
+}
+
+type DynatraceResult struct {
+	MetricId string          `json:"metricId"`
+	Data     []DynatraceData `json:"data"`
+}
+
+type DynatraceData struct {
+	Timestamps []int64    `json:"timestamps"`
+	Values     []*float64 `json:"values"`
+}
+
+func (d *KeptnDynatraceProvider) EvaluateQuery(ctx context.Context, objective klcv1alpha1.Objective, provider klcv1alpha1.KeptnEvaluationProvider) (string, error) {
+	qURL := provider.Spec.TargetServer + "/api/v2/metrics/query?metricSelector=" + objective.Query
+
+	d.Log.Info("Running query: " + qURL)
+	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	defer cancel()
+	req, err := http.NewRequestWithContext(ctx, "GET", qURL, nil)
+	if err != nil {
+		d.Log.Error(err, "Error while creating request")
+		return "", err
+	}
+
+	token, err := d.getDTApiToken(ctx, provider)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Set("Authorization", "Api-Token "+token)
+	res, err := d.client.Do(req)
+	if err != nil {
+		d.Log.Error(err, "Error while creating request")
+		return "", err
+	}
+	defer func() {
+		err := res.Body.Close()
+		if err != nil {
+			d.Log.Error(err, "Could not close request body")
+		}
+	}()
+
+	// we fail later while unmarshalling
+	b, _ := io.ReadAll(res.Body)
+	result := DynatraceResponse{}
+	err = json.Unmarshal(b, &result)
+	if err != nil {
+		d.Log.Error(err, "Error while parsing response")
+		return "", err
+	}
+
+	return fmt.Sprintf("%f", d.getSingleValue(result)), nil
+}
+
+func (d *KeptnDynatraceProvider) getSingleValue(result DynatraceResponse) float64 {
+	var sum float64 = 0
+	var count int64 = 0
+	for _, r := range result.Result {
+		for _, points := range r.Data {
+			for _, v := range points.Values {
+				if v != nil {
+					sum += *v
+					count++
+				}
+			}
+		}
+	}
+	if count < 1 {
+		return 0
+	}
+	return sum / float64(count)
+}
+
+func (d *KeptnDynatraceProvider) getDTApiToken(ctx context.Context, provider klcv1alpha1.KeptnEvaluationProvider) (string, error) {
+	if provider.Spec.SecretKeyRef == (corev1.SecretKeySelector{}) {
```

Comment:
pls introduce a method for provider to have such usage

```suggestion
	if provider.IsSecretKeyRefSet() {
```

Reply:
I don't follow the unit test part. This branch check is covered by the `TestEvaluateQuery_MissingSecret` test